src/protocol-parser/protocol-parser.ts

```diff
@@ -744,6 +746,68 @@ export namespace BrowsingContext {
     };
   };
 
+  export type PrintCommand = {
+    method: 'browsingContext.print';
+    params: PrintParameters;
+  };
+
+  // All units are in cm.
+  const PrintPageSchema = zod.object({
+    height: zod.number().min(0.0).default(27.94).optional(),
+    width: zod.number().min(0.0).default(21.59).optional(),
+  });
+
+  // All units are in cm.
+  const PrintMarginSchema = zod.object({
+    bottom: zod.number().min(0.0).default(1.0).optional(),
+    left: zod.number().min(0.0).default(1.0).optional(),
+    right: zod.number().min(0.0).default(1.0).optional(),
+    top: zod.number().min(0.0).default(1.0).optional(),
+  });
+
+  /**
+   * 1-based index, e.g. 1, '5', '2-4', '6-', '-3'
+   * @see https://w3c.github.io/webdriver/#dfn-parse-a-page-range
+   */
+  const PrintPageRangesSchema = zod
+    .array(
+      zod.union([
+        zod
+          .string()
+          .min(1)
+          .regex(/^((?:\d+)|(?:\d+[-])|(?:[-]\d+)|(?:\d+[-]\d+))$/),
+        // TODO: Use `.refine()` to check the first number is lower than the second as per the spec.
+        zod.preprocess((val) => String(val), zod.number().int().min(1)),
```

Comment:
> If range is not an integer or is less than 0, return [error](https://w3c.github.io/webdriver/#dfn-error) with [error code](https://w3c.github.io/webdriver/#dfn-error-code) [invalid argument](https://w3c.github.io/webdriver/#dfn-invalid-argument).

1. It should be non-negative.
2. Do we need a preprocessor for the number?

```suggestion
        zod.number().int().min(0),
```

Reply:
It says:

> If page ranges is not [empty](https://infra.spec.whatwg.org/#list-empty), let pages be the result of [trying](https://w3c.github.io/webdriver/#dfn-try) to [parse a page range](https://w3c.github.io/webdriver/#dfn-parse-a-page-range) with page ranges and the number of pages contained in pdf data, then remove any pages from pdf data whose **one-based index** is not contained in pages.

...on the BiDi side. But you're right, the WebDriver side says 0. I am curious what "0" would mean in this context, but regardless, that's transparent from an API point of view, we're just forwarding it to CDP. Done.

Reply:
2) The preprocessor is meant to convert everything to String, so that we can have a single logic to handle both `number` and `string` for validation.

Otherwise, in our validation code, we will need to either (i) check the type (e.g. with `typeof`) and have two branches, or (ii) cast number to string anyway.